### ISSUE TITLE
3.7.1 non sef urls ampersand broken

### ISSUE THREAD

**Issue by nordmograph** [joomla/joomla-cms#16088]
### Steps to reproduce the issue
Updated from 3.7.0. to 3.7.1
Turn SEF off in Joomla config. Although in admin menu items edit pages the menu item url is set correctly with &. In the front end & are replaced with `&amp; `breaking the variables and Itemid value, breaking modules affectations.

### Expected result
`index.php?option=com_content&view=categories&id=50&Itemid=288`

### Actual result
`index.php?option=com_content&amp;view=categories&amp;id=50& amp;Itemid=288`

### System information (as much as possible)
PHP Version	7.0.18

### Additional comments
Observed on 4 sites hosted on 2 different servers

**Comment by ot2sen** [comment]
Tested using PHP 7.0.19 and do see the replacement but without the spacing you have.
Getting URLs like index.php?option=com_content&amp;view=article&amp;id=8&amp;Itemid=116 which works.
Looks like & is kept in modules for example latest articles which doesnt replace & with &amp; in the Url generated.

**Comment by nordmograph** [comment]
No the space is only here (in this topic) so the & amp; is not displayed here as a single &. I should have used code tags.
I get 
`index.php?option=com_content&amp;view=categories&amp;id=50&amp;Itemid=288`
instead of
`index.php?option=com_content&view=categories&id=50&Itemid=288`

**Comment by PhilETaylor** [comment]
There were changes to ampReplace in Joomla 3.7.1

https://github.com/joomla/joomla-cms/commit/a7705bf1397a9a85f51a55688f743d130f325875#diff-17d077571f8b5fe34c3d9f7fec9a6e44

**Comment by PhilETaylor** [comment]
@wilsonge @mbabker The framework merge has broke this ...

**Comment by PhilETaylor** [comment]
try removing `htmlspecialchars` in the modules/mod_menu/tmpl/default_component.php file on line 52 to make that line: 

```php
echo JHtml::_('link', JFilterOutput::ampReplace($item->flink), $linktype, $attributes);
```

that fixes the issue I think

**Comment by mbabker** [comment]
Give us a unit test case on the repo to work with then.  https://github.com/joomla-framework/filter/pull/22 didn't break [the lone test case](https://github.com/joomla-framework/filter/blob/master/Tests/OutputFilterTest.php#L143) so if it can be proven that change broke it that will help us to fix it.

**Comment by PhilETaylor** [comment]
> Give us a unit test case on the repo to work with then

Actually if you check I did [actually contrib several more tests](https://github.com/joomla/joomla-cms/commit/61b9fcc4702467fdb0500057c49577037561fcc1) on this exact merge! 

Im unclear what the exact problem is here, but the framework change broke joomla, but im not sure if thats cause joomla code was crap and the framework was not...

After being shot down repeatedly by certain senior project member telling me that unit tests are *******  Im in no rush to get out of bed and write more

**Comment by PhilETaylor** [comment]
Im thinking this is joomla, not the new ampReplace, as the input to the ampReplace method is:

`index.php?option=com_content&amp;amp;view=article&amp;amp;id=1&amp;amp;Itemid=101`

which is clearly wrong, so my [suggestion here](https://github.com/joomla/joomla-cms/issues/16088#issuecomment-302194080) is probably right

**Comment by mbabker** [comment]
Note to self, merge additional tests up...

**Comment by PhilETaylor** [comment]
Please test #16089 and see if that fixes your problem

These were the only 2 cases I could see where a htmlspecialchars was running before the ampReplace

**Comment by nordmograph** [comment]
I confirm that, as stated PhilETaylor,  removing the htmlspecialchars from
modules/mod_menu/tmpl/default_component.php
and
modules/mod_menu/tmpl/default_url.php (for non absolut external url links)
fixes the issue.

**Comment by clausmunk** [comment]
Problem is also solved by simply undoing the change to OutputFilter::ampreplace

preg_replace parameters look wrong

**Comment by PhilETaylor** [comment]
> Problem is also solved by simply undoing the change to OutputFilter::ampreplace

No that is the incorrect way forward. A step back. 

The preg_replace is fine.

**Comment by franz-wohlkoenig** [comment]
closed as having PR #16089<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16088">issues.joomla.org/tracker/joomla-cms/16088</a>.</sub>
